BUG: Fixes gh-7491, pdf at x=0 of fisk/burr/burr12/f distributions.

pvanmulbregt opened the pull request:
Enable computation of pdf, logpdf at left-hand end points of `fisk`, `burr`, `burr12` and `f` distributions.  Fixes gh-7491 and half of gh-9544.
Added regression test `TestBurr` class to `stats/tests/test_distributions.py`.  `TestBurr::test_burr_stats_9544` is the same test as `test_burr_stats` in gh-9978.

pvanmulbregt commented:
There is a decent amount of overlap between gh-9983 and  gh-9978.  It may make sense to combine the two PRs.

rlucas7 commented:
@pvanmulbregt this indeed addresses some of the issues in PR #9978 (new version of #9544 after rebase fail). I'd be happy if your changes here and my changes on the other PR are combined.
If there is anything needed from me for combining the 2 PRs, please let me know.

pvanmulbregt commented:
The one test failure is in Appveyor
```
FAILED ..\..\spatial\tests\test_qhull.py::TestUtilities::test_more_barycentric_transforms
```
which seems unrelated to this PR.

pvanmulbregt commented:
@rlucas7: I grabbed most of your gh-9978 and incorporated into gh-9983.  I had a couple of questions on the remainder of the PR, and I've added some comments in a review of gh-9978.

I did have to change `burr_gen.METHOD(self,...)` in `fisk` to be `burr.METHOD(...)` as otherwise an error would occur several levels inside `burr_gen` as it expected a `self` of type `burr_gen`, not `fisk_gen`.

pvanmulbregt commented:
Rebased to get past a CI doc build failure.

rlucas7 commented on scipy/stats/_continuous_distns.py:
```diff
@@ -904,16 +973,32 @@ class fisk_gen(burr_gen):
+        return burr._logsf(x, c, 1.0)
 
     def _ppf(self, x, c):
-        return burr_gen._ppf(self, x, c, 1.0)
```
I was trying to reason through the MRO/dispatch here and was getting errors too. This looks like it resolves the dispatch.

pvanmulbregt replied:
Yes,  calling `burr_gen._ppf(self, x, c, 1.0)` with a `self` of a different type occasionally hit an error when the called method then tried to call one of its methods which required a `self` of type `burr_gen`.

rlucas7 commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -2715,6 +2726,41 @@ def test_edge_cases(self):
+class TestBurr(object):
+    def test_endpoints_7491(self):
+        # gh-7491
+        # Compute the pdf at the left endpoint dst.a.
```
There were also some changes to `_cdf` and logged versions of these methods too. Those should be tested as well in this `TestBurr` class.

pvanmulbregt replied:
Those changes are tested in other generic routines, now that the list of of x-values to be tested has bene expanded to include (finite) endpoints.

rlucas7 replied:
right, the generic tests for all distributions will get test the changes for pdf, cdf and log versions.

rlucas7 replied:
ok, that makes sense.

rlucas7 commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -2715,6 +2726,41 @@ def test_edge_cases(self):
+    def test_endpoints_7491(self):
+        # gh-7491
+        # Compute the pdf at the left endpoint dst.a.
+        data = [
```
Should there be a unit test here for the `f` distribution as well? 
It might make sense for a separate class, or if kept here change the test class name to something like `TestParameterDependentEndpoints(object):` with different tests method for the `burr` family of distributions and one for `f` endpoints (2 separate methods in the test class. Given the pecularities of the distributions this is not likely to be impacting each distribution and requires some specific domain expertise, so I'd prefer a single test class with 2 methods but I'm not firm on that if there are good arguments to push these tests out to each distribution.

pvanmulbregt replied:
The `f` has its own test of the endpoints `TestF::test_endpoints`.

rlucas7 commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -2715,6 +2726,41 @@ def test_edge_cases(self):
+        assert_allclose(mean, mean_hc)
+        assert_allclose(variance, variance_hc)
+
+    def test_burr_nan_mean_var_9544(self):
```
Question: can other methods be added (i.e. split the test the into 4) where first case tests `0<c<1` and second test does `1<=c<2`, these tests would then get the infinite mean case and infinite variance cases (respectively)? Alternately this can all be in 1 test method which fails if any of the cases fails provided a useful/informative error message comes back to stdout.

pvanmulbregt replied:
Added a check of `1<=c<2`.

rlucas7 commented on scipy/stats/_continuous_distns.py:
```diff
@@ -757,37 +757,103 @@ class burr_gen(rv_continuous):
+            lambda c, e1, e2, e3, mu2_if_c: (e3 - 3*e2*e1 + 2*e1**3) / np.sqrt((mu2_if_c)**3),
+            fillvalue=np.nan)
+        g2 = _lazywhere(
+            c > 3.0,
```
Here I think you need `c>4.0`, right?

pvanmulbregt replied:
Yes, changed to 4.0.

pvanmulbregt commented on scipy/stats/_continuous_distns.py:
```diff
@@ -757,37 +757,103 @@ class burr_gen(rv_continuous):
     """
-    _support_mask = rv_continuous._open_support_mask
+    # Do not set _support_mask to rv_continuous._open_support_mask
+    # Whether the left-hand endpoint is suitable for pdf evaluation is dependent
```
Done.

rlucas7 replied:
Sorry I wasn't clearly describing my recommendation with this comment. What I am suggesting is to specify in the docstring for which values of `c` and `d` the `x=0` point is defined and for which it is not. I see the description `x \in [0, \infty)` (or the docstring equivalent) but I'm not seeing for which values of `c,d` the value `x=0.0` is permitted? Maybe did I misunderstand the edge case for `x=0.0`?

pvanmulbregt replied:
Added statement to the doctoring for `burr_gen`.

rlucas7 replied:
Did you push the docstring changes? I'm not seeing it in the PR.

pvanmulbregt replied:
Hmm. It should be visible. 
https://github.com/scipy/scipy/blob/bec630e06142c366a54ff242b5e880c932878ecb/scipy/stats/_continuous_distns.py#L771

pvanmulbregt commented on scipy/stats/_continuous_distns.py:
```diff
@@ -757,37 +757,103 @@ class burr_gen(rv_continuous):
+        if np.any(~cond):
+            xnz = xx[~cond]
+            output[~cond] = c * d * (xnz ** (-c - 1.0)) / ((1 + xnz ** (-c)) ** (d + 1.0))
+        if output.ndim == 0:
```
The whole body has been reworked and simplified using a `_lazywhere`.

rlucas7 replied:
I like the `_lazywhere`  change, it simplified the code some too. This might be a bit tricky for someone to comprehend though but let's leave it as your changes unless someone else also thinks is less readable.

pvanmulbregt commented on scipy/stats/_continuous_distns.py:
```diff
@@ -757,37 +757,103 @@ class burr_gen(rv_continuous):
+            return output[()]
+        return output
+
+    def _logpdf(self, x, c, d):
```
Not sure about the "numerical precision". Why is the `_logpdf` different from/better than the `_pdf`?

rlucas7 replied:
with `_logpdf` you are working in "log space" so you retain more bits of precision and only move out of log-space and back into the `_pdf` space at the end. I'm not sure if it will practically make much difference in this case for most values of `c` and `d` but for some values it might matter, plus it would remove the redundant code that is shared across `_pdf` and `_logpdf`

pvanmulbregt replied:
Can you explain further why working in "log space" retains more precision in this case?

pvanmulbregt replied:
I don't see why `sc.xlog1py` would be always more precise.  For some values of `x, c, d`, `(1 + x ** (-c)) ** (d + 1.0)`can be done exactly whereas computing with logs doesn't guarantee preserving exactness.  Additionally `z -> exp(z)` has a relative condition number of `|z|`, so any extra precision due to computing in log-space can be lost when converting back to probability space.  An implementation of `a**b` might choose to do the operation in log-space, but it is not required to.

pvanmulbregt commented on scipy/stats/_continuous_distns.py:
```diff
@@ -834,10 +900,13 @@ class burr12_gen(rv_continuous):
 
     """
-    _support_mask = rv_continuous._open_support_mask
+    # _support_mask = rv_continuous._open_support_mask
```
Deleted.

pvanmulbregt commented on scipy/stats/_continuous_distns.py:
```diff
@@ -757,37 +757,103 @@ class burr_gen(rv_continuous):
+        return output
 
     def _cdf(self, x, c, d):
         return (1 + x**(-c))**(-d)
```
It could but it doesn't need to. `x=0` is already special cased by the parent `cdf` function so `x**(-c)` is always computed with a non-zero `x`.

rlucas7 replied:
>x=0 is already special cased by the parent cdf function so x**(-c) is always computed with a non-zero x.

Can you elaborate? I'm not sure I follow.

pvanmulbregt replied:
`rv_continuous.cdf(x,...)` only calls `<DISTN>._cdf(x, ...)` for `x` values inside the support of the distribution.  First it calls `self._open_support_mask(x)` and combines that with `x <= self.b` (The latter check appears to be redundant and may have predated the use of `_open_support_mask`t.) So `x=0` is never passed to `burr._cdf()`,`burr._logcdf()`.

rlucas7 commented on scipy/stats/tests/test_continuous_basic.py:
```diff
@@ -358,20 +358,28 @@ def check_pdf(distfn, arg, msg):
+    with suppress_warnings() as sup:
+        sup.filter(category=RuntimeWarning, message="divide by zero encountered in true_divide")
+        sup.filter(category=RuntimeWarning, message="divide by zero encountered in log")
+        sup.filter(category=RuntimeWarning, message="divide by zero encountered in power")  # gengamma
```
What is the impact of suppressing these warnings? 

Would it be safer to implement a list that excludes the `x=0.0` case from certain distributions? 

That seems to be a common strategy used in other unit tests, e.g. complex derivative unit tests, `expect()` method is unstable for infinite skewness and kurtosis etc. 

I am not sure if the use of exclusion/inclusion lists in unit tests is something we would like to promote or curb in scipy but I'd rather have the warnings go through in all cases except for distributions where we've explicitly decided to suppress.

pvanmulbregt replied:
I reverted `check_pdf_logpdf` to its original implementation and added `check_pdf_logpdf_at_endpoints` to check the endpoints. It isn't surprising that the computation runs into issues, as the pdf could easily be 0 (or +infinity) at the endpoints. 
Calling `distfn._support_mask(vals)` probably should have removed the need for the warning suppression, but it didn't catch all the problematic cases.

pvanmulbregt commented:
Is there an issue with Travis CI and  scipy/optimize/_trustregion_constr/tests/test_canonical_constraint.py::test_initial_constraints_as_canonical?  I changed a docstring and a `stats` testing function in the last commit and suddenly this `optimize` error showed up in the Travis CI tests.

WarrenWeckesser commented:
> Is there an issue with Travis CI ...

Yes: https://github.com/scipy/scipy/issues/9308

pvanmulbregt commented:
Thanks.  Apparently it is not consistent which explains why it could appear now.

pvanmulbregt commented:
The Travis failures are unrelated:
```
FAILED scipy/optimize/_trustregion_constr/tests/test_canonical_constraint.py::test_initial_constraints_as_canonical
```

Likely fixed in gh-10046.

pvanmulbregt commented:
Rebased and force-pushed to pick up gh-10046.

chrisb83 commented:
Note: Dagum is stats.mielke, Cross-ref #10154

chrisb83 commented on scipy/stats/_continuous_distns.py:
```diff
@@ -757,37 +757,93 @@ class burr_gen(rv_continuous):
+                ((e4 - 4*e3*e1 + 6*e2*e1**2 - 3*e1**4) / mu2_if_c**2) - 3),
+            fillvalue=np.nan)
+        return mu, mu2, g1, g2
+
```
I think _munp only works if n < c  (the expression for moments is given in the Dagum guide of Kleiber)

chrisb83 commented:
I think I also need to fix the pdf at x=0 for mielke in #10154. maybe it even makes sense to treat `mielke` as special case of `burr` and link all the methods?

pvanmulbregt commented:
The doc build failure is the `scipy/integrate/_quad_vec.py` `multiprocessing.Pool` reference fixed in 
gh-10203.

pvanmulbregt commented:
Closing in favor of gh-10381 as rebase had trouble merging with gh-10154.
